File: go-web-app/handlers/userHandler.go
package handlers

import (
    "encoding/json"
    "net/http"
    "go-web-app/services"
	"go-web-app/models"
    "github.com/gorilla/mux"
)

func GetUserHandler(w http.ResponseWriter, r *http.Request) {
    vars := mux.Vars(r)
    userId := vars["id"]

    user, err := services.GetUser(userId)
    if err != nil {
        http.Error(w, err.Error(), http.StatusNotFound)
        return
    }

    json.NewEncoder(w).Encode(user)
}

func CreateUserHandler(w http.ResponseWriter, r *http.Request) {
    var user models.User
    err := json.NewDecoder(r.Body).Decode(&user)
    if err != nil {
        http.Error(w, err.Error(), http.StatusBadRequest)
        return
    }

    if user.ID == "" || user.Name == "" {
        http.Error(w, "Invalid input: ID and Name are required", http.StatusBadRequest)
        return
    }

    err = services.CreateUser(&user)
    if err != nil {
        http.Error(w, "Failed to create user: "+err.Error(), http.StatusInternalServerError)
        return
    }

    // レスポンスヘッダーとステータスコードの設定
    w.Header().Set("Content-Type", "application/json")
    w.WriteHeader(http.StatusCreated)
    json.NewEncoder(w).Encode(user)
}

func UpdateUserHandler(w http.ResponseWriter, r *http.Request) {
    vars := mux.Vars(r)
    userId := vars["id"]

    var requestData struct {
        Name string `json:"name"`
    }

    err := json.NewDecoder(r.Body).Decode(&requestData)
    if err != nil {
        http.Error(w, "Invalid request body", http.StatusBadRequest)
        return
    }
    if requestData.Name == "" {
        http.Error(w, "Name is required", http.StatusBadRequest)
        return
    }

    updatedUser, err := services.UpdateUser(userId, requestData.Name)
    if err != nil {
        http.Error(w, "Failed to update user: "+err.Error(), http.StatusInternalServerError)
        return
    }

    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(updatedUser)
}

File: rust-web-app/src/handlers/user_handler.rs
// src/handlers/user_handler.rs

use actix_web::{web, HttpResponse};
use serde_json::json;

use crate::services::user_service;
use crate::models::user::User;

// ユーザーを取得するハンドラー
pub async fn get_user(id: web::Path<String>) -> HttpResponse {
    // パスからユーザーIDを取得
    let user_id = id.into_inner();
    
    // サービス層でユーザーを取得
    match user_service::get_user(&user_id).await {
        Ok(user) => HttpResponse::Ok().json(user), // 成功した場合、ユーザー情報をJSONとして返す
        Err(_) => HttpResponse::NotFound().json(json!({"error": "User not found"})), // ユーザーが見つからない場合、404エラーを返す
    }
}

// ユーザーを作成するハンドラー
pub async fn create_user(user: web::Json<User>) -> HttpResponse {
    let user_data = user.into_inner();
    
    // サービス層でユーザーを作成
    match user_service::create_user(user_data).await {
        Ok(user) => HttpResponse::Created().json(user), // 成功した場合、作成したユーザー情報をJSONとして返す
        Err(_) => HttpResponse::InternalServerError().json(json!({"error": "Failed to create user"})), // 失敗した場合、500エラーを返す
    }
}

// ユーザーを更新するハンドラー
pub async fn update_user(id: web::Path<String>, user: web::Json<User>) -> HttpResponse {
    let user_id = id.into_inner();
    let new_data = user.into_inner();
    
    // サービス層でユーザーを更新
    match user_service::update_user(&user_id, new_data.name).await {
        Ok(user) => HttpResponse::Ok().json(user), // 成功した場合、更新後のユーザー情報をJSONとして返す
        Err(_) => HttpResponse::InternalServerError().json(json!({"error": "Failed to update user"})), // 失敗した場合、500エラーを返す
    }
}
